### PR TITLE
Feat/lean 105 implement post method to tag entities

### PR DESCRIPTION
**Pull request by nizami**

### CODE REVIEW

File: Leanda.Categories/Domain/Aggregates/EntityCategories.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        /// User Id of person who changed the file last time
+        /// </summary>
+        public Guid UpdatedBy { get; protected set; }
+
+        /// <summary>
+        /// Date and time when file was changed last time
+        /// </summary>
+        public DateTimeOffset UpdatedDateTime { get; protected set; }
+
+        public List<Guid> CategoriesIds { get; protected set; }
```
Comment (AleksanderSidorenko) [inline Leanda.Categories/Domain/Aggregates/EntityCategories.cs]: please declare all generic collections as IList or IEnumerable.

File: Leanda.Categories/Domain/Aggregates/EntityCategories.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        private void Apply(EntityCategoriesCreated e)
+        {
+            CreatedBy = e.UserId;
+            CreatedDateTime = e.TimeStamp;
+            UpdatedBy = e.UserId;
+            UpdatedDateTime = e.TimeStamp;
+            CategoriesIds = e.CategoriesIds;
+        }
```
Comment (AleksanderSidorenko) [inline Leanda.Categories/Domain/Aggregates/EntityCategories.cs]: initialize list of category ids here as new List

File: Leanda.Categories/Domain/Commands/AddCategoriesToEntity.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿using MassTransit;
+using System;
+using System.Collections.Generic;
+
+namespace Leanda.Categories.Domain.Commands
+{
+    public interface AddCategoriesToEntity : CorrelatedBy<Guid>
+    {
+        Guid Id { get; }
+        List<Guid> CategoriesIds { get; set; }
```
Comment (AleksanderSidorenko) [inline Leanda.Categories/Domain/Commands/AddCategoriesToEntity.cs]: IEnumerable here

File: Leanda.Categories/Domain/Events/EntityCategoriesAddPersisted.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿using CQRSlite.Events;
+
+namespace Leanda.Categories.Domain.Events
+{
+    public class EntityCategoriesAddPersisted
```
Comment (AleksanderSidorenko) [inline Leanda.Categories/Domain/Events/EntityCategoriesAddPersisted.cs]: declare all event models as interfaces

File: Leanda.Categories/Domain/Events/EntityCategoriesCreated.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+﻿using Leanda.Categories.Domain.ValueObjects;
+using Sds.CqrsLite.Events;
+using System;
+using System.Collections.Generic;
+
+namespace Leanda.Categories.Domain.Events
+{
+    public class EntityCategoriesCreated : IUserEvent
```
Comment (AleksanderSidorenko) [inline Leanda.Categories/Domain/Events/EntityCategoriesCreated.cs]: same here: interface

File: Leanda.Categories/Domain/Events/EntityCategoriesCreated.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	{
+        public List<Guid> CategoriesIds { get; set; }
```
Comment (AleksanderSidorenko) [inline Leanda.Categories/Domain/Events/EntityCategoriesCreated.cs]: IEnumerable

File: Sds.Osdr.WebApi.IntegrationTests/Tests/CategoryEntities/AddCategoriesToEntity.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+using Xunit.Abstractions;
+
+namespace Sds.Osdr.WebApi.IntegrationTests
+{
+    public class AddCategoriesToEntityFixture
+    {
+        public Guid CategoryId;
+        public Guid BlobId { get; set; }
+        public Guid FileId { get; set; }
+
```
Comment (AleksanderSidorenko) [inline Sds.Osdr.WebApi.IntegrationTests/Tests/CategoryEntities/AddCategoriesToEntity.cs]: too mauch white spaces here, also 57-58 and 74-75 decrease number of new lines to one

File: Leanda.Categories/Domain/Commands/AddCategoriesToEntity.cs (hunk 3 above)
Comment (AleksanderSidorenko) [inline Leanda.Categories/Domain/Commands/AddCategoriesToEntity.cs]: set? :)

File: Sds.Osdr.WebApi.IntegrationTests/Tests/CategoryEntities/AddCategoriesToEntity.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Sds.Osdr.WebApi.IntegrationTests
+{
+    public class AddCategoriesToEntityFixture
+    {
+        public Guid CategoryId;
```
Comment (AleksanderSidorenko) [inline Sds.Osdr.WebApi.IntegrationTests/Tests/CategoryEntities/AddCategoriesToEntity.cs]: any access modifier here?

File: Sds.Osdr.WebApi.IntegrationTests/Tests/CategoryEntities/AddCategoriesToEntity.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            BlobId = harness.JohnBlobStorageClient.AddResource(harness.JohnId.ToString(), "Chemical-diagram.png", new Dictionary<string, object>() { { "parentId", harness.JohnId } }).Result;
+
+            FileId = harness.WaitWhileFileProcessed(BlobId);
+        }
+    }
+
+    [Collection("OSDR Test Harness")]
+    public class AddCategoriesToEntity : OsdrWebTest, IClassFixture<AddCategoriesToEntityFixture>
+    {
+        private Guid CategoryId;
```
Comment (AleksanderSidorenko) [inline Sds.Osdr.WebApi.IntegrationTests/Tests/CategoryEntities/AddCategoriesToEntity.cs]: same here

File: Sds.Osdr.WebApi/Controllers/CategoryEntitiesController.cs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+using System.Linq;
+using Newtonsoft.Json;
+using System.Dynamic;
+using Sds.Osdr.WebApi.Extensions;
+
+namespace Sds.Osdr.WebApi.Controllers
+{
+    [Route("api/[controller]")]
+    [Authorize]
+    [UserInfoRequired]
```
Comment (AleksanderSidorenko) [inline Sds.Osdr.WebApi/Controllers/CategoryEntitiesController.cs:27]: Have doubts do we really need UserInfoRequired attribute here? @yeskela ?

**Review by AleksanderSidorenko — COMMENTED** [review]
Generally nice work @nizami-dev but also look at check status in Travis for your branch

File: Leanda.Categories/Domain/Aggregates/EntityCategories.cs (hunk 2 above)
Comment (nizami) [reply Leanda.Categories/Domain/Aggregates/EntityCategories.cs]: Why? I got it from `e.CategoriesIds` property.

File: Leanda.Categories/Domain/Aggregates/EntityCategories.cs (hunk 1 above)
Comment (nizami) [reply Leanda.Categories/Domain/Aggregates/EntityCategories.cs]: Fixed

File: Leanda.Categories/Domain/Commands/AddCategoriesToEntity.cs (hunk 3 above)
Comment (nizami) [reply Leanda.Categories/Domain/Commands/AddCategoriesToEntity.cs]: Fixed

File: Leanda.Categories/Domain/Events/EntityCategoriesAddPersisted.cs (hunk 4 above)
Comment (nizami) [reply Leanda.Categories/Domain/Events/EntityCategoriesAddPersisted.cs]: Fixed

File: Leanda.Categories/Domain/Events/EntityCategoriesCreated.cs (hunk 5 above)
Comment (nizami) [reply Leanda.Categories/Domain/Events/EntityCategoriesCreated.cs]: In this class we have a constructor. I can't use an interface.

File: Leanda.Categories/Domain/Events/EntityCategoriesCreated.cs (hunk 6 above)
Comment (nizami) [reply Leanda.Categories/Domain/Events/EntityCategoriesCreated.cs]: Fixed

File: Sds.Osdr.WebApi.IntegrationTests/Tests/CategoryEntities/AddCategoriesToEntity.cs (hunk 8 above)
Comment (nizami) [reply Sds.Osdr.WebApi.IntegrationTests/Tests/CategoryEntities/AddCategoriesToEntity.cs]: Can you provide for what property name and what modifier I should use?

File: Leanda.Categories/Domain/Aggregates/EntityCategories.cs (hunk 2 above)
Comment (AleksanderSidorenko) [reply Leanda.Categories/Domain/Aggregates/EntityCategories.cs]: Oh my fault, missed that.

File: Sds.Osdr.WebApi.IntegrationTests/Tests/CategoryEntities/AddCategoriesToEntity.cs (hunk 8 above)
Comment (AleksanderSidorenko) [reply Sds.Osdr.WebApi.IntegrationTests/Tests/CategoryEntities/AddCategoriesToEntity.cs]: public Guid CategoryId { get; private set; }

File: Leanda.Categories/Domain/Commands/AddCategoriesToEntity.cs (hunk 3 above)
Comment (AleksanderSidorenko) [reply Leanda.Categories/Domain/Commands/AddCategoriesToEntity.cs]: remove it here.

**Review by AleksanderSidorenko — APPROVED** [review]
Approved but please make sure that build does not fail

**Review by AleksanderSidorenko — APPROVED** [review]

**Review by yeskela — APPROVED** [review]

**Pull request merged**